src/component/autocomplete.js

```diff
@@ -0,0 +1,312 @@
+  //the string to be used to do a contains search with. By default it will look
+  //if the name contains this value
+  @bindable search = '';
+
+  //can be used to select default element visually
+  @bindable selected;
+
+  //the property to query on.
+  @bindable attribute    = 'name';
+
+  //used to pass the "selected" value to the user's view model
+  @bindable({defaultBindingMode: bindingMode.twoWay}) value = null;
+
+  //the results returned from the endpoint. These can be observed and
+  //mutated.
+  @bindable({defaultBindingMode: bindingMode.twoWay}) results = [];
+
+  //used to determine the string to be shown as option label
+  @bindable label        = result => result[this.attribute];
+
+  // allow to overwrite the default apiEndpoint
+  @bindable endpoint;
+
+  // sort method that takes a list and returns a sorted list. No sorting by
+  // default.
+  @bindable sort = items => items;
+
+  // used to make the criteria more specific
+  @bindable criteria = {};
+
+  constructor(dom, api, element, queue) {
+    this.queue       = queue;
+    this.element     = element;
+    this.dom         = dom;
+    this.apiEndpoint = api;
+  }
+
+  bind() {
+    if (!this.resource) {
+      return logger.error('auto complete requires resource to be defined');
+    }
+
+    this.apiEndpoint = this.apiEndpoint.getEndpoint(this.endpoint);
+  }
+
+  /**
+   * converts a human readable string to a event keyCode
+   *
+   * @param {string} keyName a human readable key representation
+   *
+   * @returns {number} which matches the keyCode
+   */
+  keyCodes = {
+    down : 40,
+    up   : 38,
+    enter: 13,
+    tab  : 9,
+    '*'  : '*'
+  }
+
+  /**
+   * registers a event listener for the keydown
+   *
+   * @param {Element} element dom element
+   * @param {string}  keyName human readable key name
+   * @param {function} callback to be called when event is triggered
+   */
+  registerKeyDown(element, keyName, eventCallback) {
+    let eventFunction = event => {
+      if (this.keyCodes[keyName] === event.keyCode || keyName === '*') {
+        eventCallback(event);
+      }
+    };
+
+    this.listeners.push({
+      element  : element,
+      callback : eventCallback,
+      eventName: 'keydown'
+    });
+
+    element.addEventListener('keydown', eventFunction);
+  }
+
+  detached() {
+    this.removeEventListeners(this.listeners);
+  }
+
+  /**
+   * removes event listeners from DOM
+   *
+   * @param {Object[]} listeners objects that represent a event listener
+   */
+  removeEventListeners(listeners) {
+    listeners.forEach(listener => {
+      listener.element.removeEventListener(listener.eventName, listener.callback);
+    });
+  }
+
+  /**
+   * returns HTML that wraps matching substrings with strong tags
+   *
+   * @param {Object} result
+   *
+   * @returns {String}
+   */
+  labelWithMatches(result) {
+    return this.label(result).replace(
+      new RegExp(this.search, 'gi'),
+      match => {
+        return `<strong>${match}</strong>`;
```

Comment:
strong can be styled too. `strong { font-weight: normal; background-color: 'yellow';}`

I do agree that it might be nice to not do that. In that case overwriting this method is a solution.

```js
AutoCompleteCustomElement.prototype.labelWithMatches = (result) => {
  // my custom html for displaying a result.
};
```